src/librustc_typeck/impl_wf_check/min_specialization.rs

```diff
@@ -0,0 +1,315 @@
+        if let ty::Predicate::Projection(proj) = predicate {
+            let projection_ty = proj.skip_binder().projection_ty;
+            let projected_ty = proj.skip_binder().ty;
+            seen_params.extend(cgp::parameters_for(&projection_ty, true));
+            for param in cgp::parameters_for(&projected_ty, false) {
+                if !seen_params.contains(&param) {
+                    constrained_params.insert(param.0);
+                }
+            }
+            seen_params.extend(cgp::parameters_for(&projected_ty, true));
+        }
+    }
+
+    impl_substs
+        .iter()
+        .enumerate()
+        .filter(|(idx, _)| !constrained_params.contains(&(*idx as u32)))
+        .map(|(_, arg)| *arg)
+        .collect()
+}
+
+fn check_duplicate_params<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    impl1_substs: SubstsRef<'tcx>,
+    parent_substs: &Vec<GenericArg<'tcx>>,
+    span: Span,
+) {
+    let mut base_params = cgp::parameters_for(parent_substs, true);
+    base_params.sort_by_key(|param| param.0);
+    if let (_, [duplicate, ..]) = base_params.partition_dedup() {
+        let param = impl1_substs[duplicate.0 as usize];
+        tcx.sess
+            .struct_span_err(span, &format!("specializing impl repeats parameter `{}`", param))
+            .emit();
+    }
+}
+
+fn check_static_lifetimes<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    parent_substs: &Vec<GenericArg<'tcx>>,
+    span: Span,
+) {
+    if tcx.any_free_region_meets(parent_substs, |r| *r == ty::ReStatic) {
+        tcx.sess.struct_span_err(span, &format!("cannot specialize on `'static` lifetime")).emit();
+    }
+}
+
+fn check_predicates<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    impl1_def_id: DefId,
+    impl1_substs: SubstsRef<'tcx>,
+    impl2_node: Node,
+    impl2_substs: SubstsRef<'tcx>,
+    span: Span,
+) {
+    let impl1_predicates = tcx.predicates_of(impl1_def_id).instantiate(tcx, impl1_substs);
+    let mut impl2_predicates = if impl2_node.is_from_trait() {
+        // Always applicable traits have to be always applicable without any
+        // assumptions.
+        InstantiatedPredicates::empty()
+    } else {
+        tcx.predicates_of(impl2_node.def_id()).instantiate(tcx, impl2_substs)
+    };
+    debug!(
+        "check_always_applicable(\nimpl1_predicates={:?},\nimpl2_predicates={:?}\n)",
+        impl1_predicates, impl2_predicates,
+    );
+
+    // Since impls of always applicable traits don't get to assume anything, we
+    // can also assume their supertraits apply.
+    let always_applicable_traits: Vec<_> = impl1_predicates
+        .predicates
+        .iter()
+        .filter(|predicate| {
+            matches!(
+                trait_predicate_kind(tcx, predicate),
+                Some(TraitSpecializationKind::AlwaysApplicable)
+            )
+        })
+        .copied()
+        .collect();
+    impl2_predicates.predicates.extend(traits::elaborate_predicates(tcx, always_applicable_traits));
+
+    for predicate in impl1_predicates.predicates {
```

Comment:
So what would happen in a case like

```rust
trait Foo { }
struct MyVec<T: Eq> { t: T }
impl<T> Foo for T { }
impl<U: Eq> Foo for MyVec<U> { }
```

This is going to give an error, right? But we could accept it, I believe. We would want to consider the "WF-predicates" for the input types to impl1, basically.

Reply:
We could, it just doesn't occur in the standard library because bounds are rarely found on `libstd` types, rather than their impls.

Reply:
Yes -- but it would be really annoying in practice for types that *do* use bounds, since there would be no way to write impls like the one above. I think it's worth accepting, myself.

Reply:
This is implemented now.